### ISSUE TITLE
Operate attestations

### ISSUE THREAD

**Issue by hwwhww** [ethereum/py-evm#1616]
### What is wrong?

There *were* some validation helpers in [`eth/beacon/state_machines/validation.py`](https://github.com/ethereum/py-evm/blob/1dabf4657ad78bfbd40bd29adb560197bb839616/eth/beacon/state_machines/validation.py). Now we need to implement the new ones for the latest spec. Start from `Attestion` validation.

### How can it be fixed
Implement `operate_attestations` based on [the spec - operating attestations](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#attestations-1):

```python
def operate_attestations (state: BeaconChain, attestations: Sequence[Attestation]) -> BeaconState:
    validate_attestations(state, attestations)
    state = update_latest_attestations(state, attesations)
    return state
```

1. Implement the `validate_attestations` function in `eth/beacon/state_machines/validation.py`
    * Implement the "Verify ..." logic from spec.
    * Raise `ValidationError` exception if the check fails.
2. Implement an immutable function `update_latest_attestations` for the last line - Append `PendingAttestationRecord(data=attestation.data, participation_bitfield=attestation.participation_bitfield, custody_bitfield=attestation.custody_bitfield, slot_included=state.slot)` to `state.latest_attestations`.
    * (STUB) Implement it in `eth/beacon/state_machines/state_update.py`.  <-- may need a better name...

**Comment by djrtwo** [comment]
I think I'm in favor of using `process_` to describe what we do to the beacon chain operations. "operate ___" reads a bit strange.

- `process_attestations`
- `process_exits`
- etc...

**Comment by djrtwo** [comment]
Where should these new `operations` live? maybe `eth/beacon/state_machines/operations.py`?

I'm thinking thinking that most of the work will happen in a file like `eth/beacon/state_machines/state_transition.py` that has the following functions (or something similar):
* `state_transition`
* `per_slot_transition`
* `per_block_transition`
* `per_epoch_transition`

The operation functions (`process_attestations`, `process_exits`, etc) could live in there or be imported from some specific `operations.py` file.

**Comment by hwwhww** [comment]
Having `operations.py` for the operation functions sounds good to me. 👍 

The layout (`eth/beacon/state_machines/state_transition.py`) above makes sense. So we can follow the spec tightly at the beginning. (And yay we have better spec structure now!), then the `eth/beacon/state_machines/base.py` would be lighter (yay no more `CrystallizedState` now!).

p.s. and IMO regarding to #1502, the next thing to do may be building `per_slot_transition` and `per_block_transition`, not finishing all operations.

**Comment by hwwhww** [comment]
Another consideration:
1. Only write the abstract interfaces in `eth/beacon/state_machines/state_transition.py`.
2. Implement the main logic in `eth/beacon/state_machines/forks/serenity/state_transition.py`
3. Add `config` in `tests/beacon/conftest.py`.

Then we can use `eth/beacon/state_machines/forks/serenity/configs.py` more smoothly. Just set `config: BeaconConfig` as one of the parameters of the state transition functions.

**Comment by djrtwo** [comment]
absolutely agree with #1502. 

aah, as for the abstract interface, are you saying `state_transition.py` should house a class? In that case should we call it a `StateMachine`? I need to go brush up on the virtual machine structure in eth1.0

How do we want to handle validation wrt forks? Right now I have the validation living in `eth/beacon/state_machines/validation.py`, but a couple of the functions we already _know_ will change in phase 1.

**Comment by djrtwo** [comment]
We can talk to @cburgdorf on the call today!

**Comment by cburgdorf** [comment]
Let me add that we are trying to migrate away from the inheritance heavy approach that we use today more towards a compositional approach. But I'm not sure if we want to consider this for this specific part or if we should just stick to the eth1 system for now.

I would say that there should be a `eth/beacon/state_machines/base.py` that hosts a `BaseStateMachine`  that has all the methods `abstract`.

E.g.

```python
class BaseStateMachine:

    @staticmethod
    @abstractmethod
    def validate_something(x: int) -> None:
        pass
```

Then inside `eth/beacon/state_machines/forks/serenity/__init__.py` we'd have `SerenityStateMachine` that implements the concrete class.

```python
class SerenityStateMachine(BaseStateMachine):

    @staticmethod
    def validate_something(x: int) -> None:
        print("validating...")
```

As you already pointed out, you *could* also put the individual functions separately and then do something like.

```python
class SerenityStateMachine(BaseStateMachine):

    validate_something = staticmethod(validate_something)
```
But unfortunately there's a problem with mypy with this construct that leads to silencing the type checker which is bad.

But...I'd argue that for static methods the argument of having these functions separately for better testability does actually not hold because you can always refer to it as `SerenityStateMachine.validate_something` **without** being forced to create an instance of `SerenityStateMachine` because the function is static anyway. Same with classmethods.

For instance methods you totally *could* have those be defined separately e.g.

```python
def validate_something(self, x: int) -> None:
    print("validating...")
```

and then do

```python
class SerenityStateMachine(BaseStateMachine):

    validate_something = validate_something
```

But I'm not so sure if it makes sense to keep these separate if they depend on some instance anyway.

**Comment by djrtwo** [comment]
closed via #1631
